packages/twoside/init.lua

```diff
@@ -65,28 +66,74 @@ function package:_init (options)
   self.class:registerPostinit(function (class)
     class:mirrorMaster(options.oddPageMaster, options.evenPageMaster)
   end)
+  self.class:registerHook("newpage", spreadHook)
   self.class:registerHook("newpage", switchPage)
 end
 
 function package:registerCommands ()
 
   self:registerCommand("open-double-page", function()
-    SILE.typesetter:leaveHmode()
+    SILE.call("open-spread", { double = false, odd = true, blank = false })
+  end)
+
+  self:registerCommand("open-spread-eject", function()
     SILE.call("supereject")
-    if self.class:oddPage() then
-      SILE.typesetter:typeset("")
-      SILE.typesetter:leaveHmode()
-      SILE.call("supereject")
+  end)
+
+  -- This is upstreamed from CaSILE. Similar to the original open-double-page,
+  -- but can disable headers and folios on blank pages and allows opening the
+  -- even side (with or without a leading blank).
+  self:registerCommand("open-spread", function (options)
+    local odd = SU.boolean(options.odd, true)
+    local double = SU.boolean(options.double, true)
+    local blank = SU.boolean(options.blank, true)
+    local optionsMet = function ()
+      return (not double or spread_counter > 1) and
+             (odd and self.class:oddPage()) or
+             (not odd and not self.class:oddPage())
     end
+    spread_counter = 0
     SILE.typesetter:leaveHmode()
+    -- Output a box, then remove it otherwise we can't prove what page new content will land on
+    SILE.call("hbox"); SILE.typesetter:leaveHmode(); table.remove(SILE.typesetter.state.nodes)
+    if spread_counter == 1 and optionsMet() then
+      SILE.typesetter.state.outputQueue = {}
+      return
+    end
+    repeat
+      if spread_counter > 0 then
+        SILE.call("hbox")
+        SILE.typesetter:leaveHmode()
+        if blank then
+          SILE.scratch.headers.skipthispage = true
```

Comment:
Hmm wait! Who's defining this? We don't have a "headers" package abstraction as we have a "folios" one (N.B. I did a quick one in my own bundle...). This comes from the book class, right? But doesn't it mean that any other book class _not even inheriting the standard book class_ has to respect some unspecified scratch thing if it wants to use this command? It does seem to me there's a serious separation of concerns issue ;-)

Reply:
We already stuff the entire contents of R/L headers in a scratch variable, I'm not concerned about a scratch variable that suppresses the same without erasing it—at least not more than I think **all** the scratch variables should be refactored to be scoped to the class/package/typesetter/whatever. Yes anything wanting to muck with the existing book stuff would need to use the save variables. In this case some other header system could ignore this just fine, but then would have to implement their own way to suppress headers in spreads if they want that feature.

Reply:
I agree on the refactoring needed, and also have implemented my own header system at least 3 times without submitting any iteration to SILE because it's easy to work up what a project needs but a lot harder to generalize robustly.